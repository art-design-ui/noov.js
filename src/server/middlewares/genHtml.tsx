import React from 'react'
import { StaticRouter } from 'react-router-dom'
import { renderToString } from 'react-dom/server'
import routeList from '../../client/main/route-config'
import { Provider } from 'react-redux'
import StyleContext from 'isomorphic-style-loader/StyleContext'
import getStaticRoutes from '../utils/get-static-routes'
import matchRoute from '../../common/match-route'
import App from '../../client/main/router'
import { Store } from 'redux'

interface IGenResult {
  html: string
  fetchResult:any
}
export default async function genHtml(
  path: string,
  insertCss: Function,
  store: Store
): Promise<IGenResult> {
  // 获得静态路由
  const staticRoutesList = await getStaticRoutes(routeList)

  // 查找到的目标路由对象
  const matchResult = await matchRoute(path, staticRoutesList)
  const { targetRoute } = matchResult

  // 进行数据预取，更新 store 内的数据
  let fetchDataFn
  let fetchResult= {}
  if (targetRoute) {
    fetchDataFn = targetRoute.component ? targetRoute.component.getInitialProps : null
    if (fetchDataFn) {
      // 更新 state
      fetchResult = await fetchDataFn({ store })
    }
  }

  // 将预取数据在这里传递过去 组内通过props.staticContext获取 IntrinsicAttributes 
  const context: any= {
    initialData: fetchResult
  }
  const html = renderToString(
    <Provider store={store}>
      <StaticRouter location={path} context={context}>
        <StyleContext.Provider value={{ insertCss }}>
          <App routeList={staticRoutesList} />
        </StyleContext.Provider>
      </StaticRouter>
    </Provider>
  )
  return {
    html,
    fetchResult
  }
}
