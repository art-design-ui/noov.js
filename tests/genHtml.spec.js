import genHtml from '../src/server/middlewares/genHtml'
import getStore from '../src/client/store/reducers'

describe('test genHtml Function', () => {
  it('should return the correct result when call genHtml', async () => {
    const store = getStore()
    const insertCss = () => {}
    const { html } = await genHtml('/404', insertCss, store)
    expect(html).toMatch(`<div class="layout-box"><h1>koa+react+ssr</h1><div><p>404页面page22</p><p>12</p></div></div>`)
    expect(html).toMatchSnapshot() // 保存快照
  })
})
