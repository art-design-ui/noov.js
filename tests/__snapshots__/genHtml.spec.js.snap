// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test genHtml Function should return the correct result when call genHtml 1`] = `"<div class=\\"layout-box\\"><h1>koa+react+ssr</h1><div><p>404页面page22</p><p>12</p></div></div>"`;
